File: query_gene_info.pl
#!/usr/bin/env perl -w

# An example script demonstrating the use of BioMart API.
# This perl API representation is only available for configuration versions >=  0.5 
# Guillaume edit 2016/09/04: useful tuto at [https://www.biostars.org/p/53241/]

use strict;
use BioMart::Initializer;
use BioMart::Query;
use BioMart::QueryRunner;

my $confFile = "/users/ldog/glouvel/.local/lib/perl/biomart-perl/conf/martURLLocation85.xml";
#
# NB: change action to 'clean' if you wish to start a fresh configuration  
# and to 'cached' if you want to skip configuration step on subsequent runs from the same registry
#

my $action='cached';
my $initializer = BioMart::Initializer->new('registryFile'=>$confFile, 'action'=>$action);
my $registry = $initializer->getRegistry;


while (my $dataset = <>) {
	print STDERR $dataset ;
	chomp $dataset ;

	my $query = BioMart::Query->new('registry'=>$registry,'virtualSchemaName'=>'default');
		
	$query->setDataset("${dataset}_gene_ensembl");
	$query->addFilter("biotype", ["protein_coding"]);
	$query->addAttribute("ensembl_gene_id");
	$query->addAttribute("ensembl_transcript_id");
	$query->addAttribute("ensembl_peptide_id");
	$query->addAttribute("chromosome_name");
	$query->addAttribute("start_position");
	$query->addAttribute("end_position");
	#$query->addAttribute("strand");
	#$query->addAttribute("band");
	$query->addAttribute("external_gene_name");
	$query->addAttribute("external_gene_source");
	$query->addAttribute("transcript_count");
	$query->addAttribute("percentage_gc_content");
	$query->addAttribute("description");

	$query->formatter("TSV");

	my $query_runner = BioMart::QueryRunner->new();
	############################## GET COUNT ############################
	# $query->count(1);
	# $query_runner->execute($query);
	# print $query_runner->getCount();
	#####################################################################


	############################## GET RESULTS ##########################
	# to obtain unique rows only
	# $query_runner->uniqueRowsOnly(1);

	$query_runner->execute($query);
	open (OUT, ">${dataset}_gene_info.tsv") ;
	$query_runner->printHeader(\*OUT);
	$query_runner->printResults(\*OUT);
	$query_runner->printFooter(\*OUT);
	close OUT ;
	#####################################################################
}

File: query_gene_info.py
#!/usr/bin/env python

import urllib2
import fileinput

xmlbase = """<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE Query>
<Query  virtualSchemaName = "default" formatter = "TSV" header = "0" uniqueRows = "0" count = "" datasetConfigVersion = "0.6" >
    <Dataset name = "{dataset}_gene_ensembl" interface = "default" >
        <Filter name = "biotype" value = "protein_coding"/>
        <Attribute name = "ensembl_gene_id" />
        <Attribute name = "percentage_gc_content" />
        <Attribute name = "start_position" />
        <Attribute name = "end_position" />
        <Attribute name = "chromosome_name" />
        <Attribute name = "{species}_homolog_ensembl_gene" />
        <Attribute name = "{species}_homolog_chromosome" />
        <Attribute name = "{species}_homolog_chrom_start" />
        <Attribute name = "{species}_homolog_chrom_end" />
        <Attribute name = "{species}_homolog_orthology_type" />
        <Attribute name = "{species}_homolog_orthology_confidence" />
        <Attribute name = "{species}_homolog_perc_id" />
        <Attribute name = "{species}_homolog_perc_id_r1" />
        <Attribute name = "{species}_homolog_dn" />
        <Attribute name = "{species}_homolog_ds" />
    </Dataset>
</Query>"""

url="http://www.ensembl.org/biomart/martservice?query="
# Ensembl 85
url="http://mar2016.archive.ensembl.org/biomart/martservice?query="
